tags.go

```diff
@@ -200,3 +206,66 @@ func (client *Client) DeleteTag(id ID) error {
 	err := client.Mutate(&m, v, WithName("TagDelete"))
 	return HandleErrors(err, m.Payload.Errors)
 }
+
+// ReconcileTags manages tags API operations for TaggableResourceInterface implementations
+//
+// Tags not in 'tagsWanted' will be deleted, new tags from 'tagsWanted' will be created. Reconciled tags are returned.
+func (client *Client) ReconcileTags(resourceType TaggableResourceInterface, tagsWanted []string) ([]Tag, error) {
+	var err error
+	var tagConnection *TagConnection
+	var assignedTags []Tag
+
+	// Get underlying resource type for type checking
+	switch resource := resourceType.(type) {
+	case *Domain:
+		tagConnection, err = resource.GetTags(client, nil)
+	case *InfrastructureResource:
+		tagConnection, err = resource.GetTags(client, nil)
+	case *Repository:
+		tagConnection, err = resource.GetTags(client, nil)
+	case *Service:
+		tagConnection, err = resource.GetTags(client, nil)
+	case *System:
+		tagConnection, err = resource.GetTags(client, nil)
+	case *Team:
+		tagConnection, err = resource.GetTags(client, nil)
+	case *User:
+		tagConnection, err = resource.GetTags(client, nil)
+	default:
+		return assignedTags, fmt.Errorf("not a taggable resource type: %s" + string(resourceType.ResourceType()))
+	}
+	if err != nil {
+		return assignedTags, err
+	}
+	if tagConnection == nil {
+		return assignedTags, fmt.Errorf("no tags found on %s with id '%s'", string(resourceType.ResourceType()), resourceType.ResourceId())
+	}
+
+	// delete tags found in resource but not listed in tagsWanted
+	for _, tag := range tagConnection.Nodes {
+		if !slices.Contains(tagsWanted, tag.Flatten()) {
+			if err := client.DeleteTag(tag.Id); err != nil {
+				return assignedTags, err
+			}
+		}
+	}
+
+	// format tags listed in Terraform config but not found in service
+	tagInput := map[string]string{}
+	for _, tag := range tagsWanted {
+		parts := strings.Split(tag, ":")
+		if len(parts) != 2 {
+			return assignedTags, fmt.Errorf("[%s] invalid tag, should be in format 'key:value' (only a single colon between the key and value, no spaces or special characters)", tag)
```

Comment:
1. I think it's odd that we check this on tag insertion but not tag deletion.
2. Since we always want key value pairs, we can just make it so that the function argument `tagsWanted` is already a map of key value pairs? No need to split and parse the string that way and we have a solid schema for what the input should look like.

Reply:
1. Tag insertion takes a slice of tags where each tag is a single string of format `key:value`. Tag deletion gets its tags from an API call where each `tag.Id` is known. These two logic streams take two different types of inputs and have different expectations.
2. We _could_ change tags from strings to a mapping. But that expands the scope of this task to require changes from all the calling functions. We could also build up the `Tag` struct with constructors that handle this.

Reply:
1. Understood

2. I still think that forcing them to use key value pairs in a `map` is the way to go but they will either have to validate the input before calling this function or after calling the function so validating it in the function still works.